scripted-diff: Rename wallet database classes

ryanofsky opened the pull request:
Scripted diff to rename some wallet classes. Motivated by discussion in https://github.com/bitcoin/bitcoin/pull/11687#discussion_r155354119

| Current          | New                 |
| ---------------- | ------------------- |
| CDBEnv           | BerkeleyEnvironment |
| CDB              | BerkeleyBatch       |
| CWalletDBWrapper | WalletDatabase      |
| CWalletDB        | WalletBatch         |

Berkeley\* classes are intended to contain BDB specific code, while Wallet\* classes are intended to be more backend-agnostic.

Also renamed associated variables:

| Current             | New             |
| ------------------- | --------------- |
| dbw                 | database        |
| pwalletdb           | batch           |
| pwalletdbEncryption | encrypted_batch |

ryanofsky committed:
scripted-diff: Rename wallet database classes

-BEGIN VERIFY SCRIPT-

sed -i 's/\<CWalletDBWrapper\>/BerkeleyDatabase/g' src/wallet/db.h src/wallet/db.cpp
sed -i '/statuses/i/** Backend-agnostic database type. */\nusing WalletDatabase = BerkeleyDatabase\;\n' src/wallet/walletdb.h
ren() { git grep -l "\<$1\>" 'src/*.cpp' 'src/*.h' ':(exclude)*dbwrapper*' test | xargs sed -i "s:\<$1\>:$2:g"; }
ren CDBEnv           BerkeleyEnvironment
ren CDB              BerkeleyBatch
ren CWalletDBWrapper WalletDatabase
ren CWalletDB        WalletBatch
ren dbw              database
ren m_dbw            m_database
ren walletdb         batch
ren pwalletdb        batch
ren pwalletdbIn      batch_in
ren wallet/batch.h   wallet/walletdb.h
ren pwalletdbEncryption encrypted_batch

-END VERIFY SCRIPT-

jnewbery commented:
Concept ACK. The current naming is very confusing. There's even a comment in walletdb.h to say so:

```
/**
 * Overview of wallet database classes:
 *
 * - CDBEnv is an environment in which the database exists (has no analog in dbwrapper.h)
 * - CWalletDBWrapper represents a wallet database (similar to CDBWrapper in dbwrapper.h)
 * - CDB is a low-level database transaction (similar to CDBBatch in dbwrapper.h)
 * - CWalletDB is a modifier object for the wallet, and encapsulates a database
 *   transaction as well as methods to act on the database (no analog in
 *   dbwrapper.h)
 *
 * The latter two are named confusingly, in contrast to what the names CDB
 * and CWalletDB suggest they are transient transaction objects and don't
 * represent the database itself.
 */
```

This should probably go in just before branching v0.16, since rebasing on top of this will be a pain.

Other changes you could consider:

- manually update that overview comment (and perhaps move to `wallet.h`?)
- rename the `wallet/db` and `wallet/walletdb` files to be more descriptive. Perhaps `wallet/berkeley_db` and `wallet/db_checkpoint`, but I'm sure we could do better than that.

I think your suggested class names are good ('checkpoint' is certainly better than 'transaction'!). I can't currently think of anything better.

ryanofsky commented:
> manually update that overview comment (and perhaps move to wallet.h?)

Added commit to fix the overview comment: 51eaec47ccb61de2d791c86b785ee4cf111c3b05. I didn't move it into `wallet.h` for now because it sounds from https://github.com/bitcoin/bitcoin/pull/11687#discussion_r155354119 like a goal is to pull database reading and writing code out of main wallet class, which would make `wallet.h` not the right place for it.

> rename the wallet/db and wallet/walletdb files to be more descriptive. Perhaps wallet/berkeley_db and wallet/db_checkpoint, but I'm sure we could do better than that.

I'd be inclined to rename `wallet/db.h` to `wallet/db-berkeley.h`, leaving room for other implementations like `wallet/db-dummy.h`, `wallet/db-sqlite.h`, `wallet/db-log.h` (#5686), etc.

I'd be inclined to rename `wallet/walletdb.h` something like `wallet/serialize.h`, `wallet/schema.h`, `wallet/kvschema.h`, `wallet/db.h`, or `wallet/db-keyvalue.h` since it's mainly responsible for key value serialization and mostly not tied to the berkeleydb backend. This would make more sense with the CWalletDB -> CDB dependency removed or reversed, though.

---

Added 1 commits 003b6190e5a3d870a5eca3195e2381ee4b93de4d -> 51eaec47ccb61de2d791c86b785ee4cf111c3b05 ([pr/wren.1](https://github.com/ryanofsky/bitcoin/commits/pr/wren.1) -> [pr/wren.2](https://github.com/ryanofsky/bitcoin/commits/pr/wren.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wren.1...pr/wren.2))

TheBlueMatt commented:
Nice! I find the "WalletCheckpoint" name kinda confusing, it is a DB batch/transaction, so I'd kinda prefer we stick with "normal" nomenclature there. I'm really not sure that "BerkeleyCheckpoint" makes all that much sense for CDB, either. I kinda wish we'd split it out into an interface and implementation, though not sure if its required for this PR, but its already essentially a pretty clean interface towards a wallet DB backend.

ryanofsky commented:
> I find the "WalletCheckpoint" name kinda confusing

Yeah, actually checkpoint is a bad name. I just chose it because I saw a `txn_checkpoint` call in `CDB::Flush` and assumed "checkpoint" was some kind of BDB synonym for transaction. But it's really not. It would be good to avoid the word transaction so maybe the best thing to call this is something like `Batch` or `Update`.

jnewbery commented:
Can you push your latest branch onto `pr/wren` so it shows up here?

ACK changing Checkpoint to Batch or Commit. Please just not transaction!

ACK your suggested filename changes. My preferences would be:
- `wallet/db.h` *->* `wallet/db-berkeley.h`
- `wallet/walletdb.h` *->* `wallet/schema.h`

I've reviewed your commit changing the overview comment. I think it could be improved further by removing the 'analogous to x in dbwrapper.h' parts, and perhaps changing the ordering and wording, eg:

```
/**
 * Overview of wallet database classes:
 *
 * - WalletBatch is an abstract modifier object for the wallet database, and encapsulates a database
 *   batch update as well as methods to act on the database. It should be agnostic to the database implementation.
 *
 * The following classes are implementation specific:
 * - BerkeleyEnvironment is an environment in which the database exists.
 * - BerkeleyDatabase represents a wallet database.
 * - BerkeleyBatch is a low-level database batch update.
 */

TheBlueMatt commented:
> It would be good to avoid the word transaction so maybe the best thing to call this is something like Batch or Update or Commit.

Good point. We use the term "Batch" in the context of LevelDB/CCoin stuff, so I'd vote for that.

ryanofsky committed:
Update walletdb comment after renaming.

Text from https://github.com/bitcoin/bitcoin/pull/11851#issuecomment-350320608
by John Newbery <john@johnnewbery.com>.

ryanofsky commented:
Updated 51eaec47ccb61de2d791c86b785ee4cf111c3b05 -> 4299018f8f1f2b952c93f79b82073bcaa96807f6 ([pr/wren.2](https://github.com/ryanofsky/bitcoin/commits/pr/wren.2) -> [pr/wren.3](https://github.com/ryanofsky/bitcoin/commits/pr/wren.3)), switching from checkpoint to batch, dropping the CWallet->Wallet rename (because it isn't really related and makes the change bigger), adding some variable renames for consistency, and adding John's comment.

The only thing I didn't do yet is rename files. There might be other suggestions for names, and file renames could also happen in a separate PR.

jnewbery commented:
I'm going to hold off reviewing until we're closer to a major release (since I think this only makes sense going in right before branching). Russ - let me know if you disagree and think we should be reviewing/merging sooner. If this does get merged just before a branch, then I think there's no issue with doing `s/CWallet/Wallet` at the same time.

ryanofsky commented:
Sounds good to me. It would also make things a little easier for me if this were merged after #11687.

TheBlueMatt commented:
I'm not sure I'd call BerkeleyBatch so low-level or recommend it be BDB-specific. Its public interface is pretty DB-agnostic - Read/Write/Verify/PeriodicFlush.

ryanofsky commented:
> I'm not sure I'd call BerkeleyBatch so low-level or recommend it be BDB-specific. Its public interface is pretty DB-agnostic - Read/Write/Verify/PeriodicFlush.

Yeah, that code could be pulled out into WalletBatch or a new key-value store generic class. But I think John's comment is basically right. WalletBatch code should be agnostic to berkeley db details and potentially able to work with another key-value backend, while BerkeleyBatch should contain more berkeley specific stuff. This is mostly true right now, and could be made more true with some refactoring, which I can follow up with. You should talk to John though about whether you prefer this comment or the previous comment, or want to combine them somehow.

practicalswift commented:
Concept ACK

laanwj commented:
Agree on the naming and using `Batch`.

However I agree with this as well:

> I'm not sure I'd call BerkeleyBatch so low-level or  recommend it be BDB-specific. Its public interface is pretty DB-agnostic  - Read/Write/Verify/PeriodicFlush.

I tend to agree. I have a local patch set that ports the wallet functions to LevelDB. I don't think the interface should hardcode the database implementation name.

If you do want to do this, add a typedef so the non-database-specific code can stay without Berkeley* references. e.g.:
```
typedef BerkeleyEnvironment WalletDBEnvironment; 
typedef BerkeleyDatabase WalletDatabase;
typedef BerkeleyBatch WalletDatabaseBatch;
```

ryanofsky commented:
@laanwj, is there a link to your leveldb patch? I'm surprised there's much code in these classes that's reusable for leveldb. BerkeleyEnvironment and BerkeleyDatabase in particular seem pretty tied to Berkeley, though BerkeleyBatch is more independent.

I'm thinking perhaps I should move the independent parts of BerkeleyBatch out into WalletBatch along with this renaming to avoid creating new confusion.

> after all we don't want to have to do another batch rename every time the database changes

I can clean this up more too, but after #11687 this might not be as much of as problem as you would think. #11687 eliminates most references to the BerkeleyEnvironment, BerkeleyBatch, and BerkeleyDatabase outside of the db files.

Anyway, I think I will mark this WIP for now because there are other improvements I can make to go along with the renaming, and I think reviewer time would be better spent on #11687 anyway.

ryanofsky commented:
I think I found the leveldb wallet branch, it is https://github.com/laanwj/bitcoin/commits/20170310_experiment_leveldb_wallet and #9951. #5686 is also similar.

laanwj commented:
> @laanwj, is there a link to your leveldb patch? I'm surprised there's much code in these classes that's reusable for leveldb.

Right - the patch replaces pretty much all the code *inside* the classes.
My only comment is about the exterior. So wallet.cpp etc should not be referring to `Berkeley` anything as class name.

ryanofsky commented:
> My only comment is about the exterior. So wallet.cpp etc should not be referring to Berkeley anything as class name

Ah, I see. Your concern was narrower than I thought and can be addressed by a typedef like you suggested. Will fix and have the PR ready again soon. I still might want to merge this after #11687, though since #11687 touches a lot of the same code and this being a scripted-diff is much easier to rebase

PierreRochard commented:
Concept ACK. Was changing `dbw` to `database` in the src/test/dbwrapper_tests.cpp file intentional? That would seem to increase the scope of this PR beyond the wallet code

ryanofsky commented:
PSA: Everybody commenting here please stop reviewing this PR and look at #11687 instead :)

> Was changing dbw to database in the src/test/dbwrapper_tests.cpp file intentional? That would seem to increase the scope of this PR beyond the wallet code

This was unintentional but seemed harmless so I just left it in. I can try to tweak the scripted-diff to prevent it.

laanwj commented:
>  Will fix and have the PR ready again soon.

Thanks. Sorry for being unclear at first.

sipa commented:
Pieter was here.

MarcoFalke commented:
Needs rebase to fix travis (sorry)

ryanofsky commented:
> Needs rebase to fix travis (sorry)

Rebased e0ad11506a62c7ad28411ba67abad68131058bf5 -> 333ede1b54f00b063a311cf8bf7d70c711fa234f ([pr/wren.10](https://github.com/ryanofsky/bitcoin/commits/pr/wren.10) -> [pr/wren.11](https://github.com/ryanofsky/bitcoin/commits/pr/wren.11))

PierreRochard commented:
Opening comment could be updated to include `WalletDatabase`/`BerkeleyDatabase` distinction as well as the variable renames:

| Current  | New |
| ------------- | ------------- |
| dbw  | database  |
| m_dbw  | m_database  |
| pwalletdb  | batch  |
| pwalletdbIn  | batch_in  |
| walletdb  | batch  |

Should pwalletdbEncryption be renamed as well?

ryanofsky commented:
> Opening comment could be updated to include WalletDatabase/BerkeleyDatabase distinction as well as the variable renames:

Thanks, updated to note the distinction and variable renames.

> Should pwalletdbEncryption be renamed as well?

Good catch. Added `ren pwalletdbEncryption encrypted_batch` line to the script.

Updated 333ede1b54f00b063a311cf8bf7d70c711fa234f -> e2ce4c00c141275fbbf733dc0abef1bb4d986065 ([pr/wren.11](https://github.com/ryanofsky/bitcoin/commits/pr/wren.11) -> [pr/wren.12](https://github.com/ryanofsky/bitcoin/commits/pr/wren.12)) with just the change to the scripted-diff.

PierreRochard commented:
Tested ACK e2ce4c00c141275fbbf733dc0abef1bb4d986065

FYI in your scripted diff there's `ren wallet/batch.h   wallet/walletdb.h` which doesn't do anything because it's in the wrong order.

ryanofsky commented:
Rebased e2ce4c00c141275fbbf733dc0abef1bb4d986065 -> b21912d455f225740fc5d4356cb4b31379c077d2 ([pr/wren.12](https://github.com/ryanofsky/bitcoin/commits/pr/wren.12) -> [pr/wren.13](https://github.com/ryanofsky/bitcoin/commits/pr/wren.13)) because of various merge conflicts.

> FYI in your scripted diff there's ren wallet/batch.h wallet/walletdb.h which doesn't do anything because it's in the wrong order.

I don't think this is true. That line is needed to prevent the earlier `ren walletdb` from mangling #includes. I think it's possible you were seeing an error from the `ren dbw_in` which is removed because it is no longer applicable after #11687.

PierreRochard commented:
> That line is needed to prevent the earlier ren walletdb from mangling #includes.

Ah I see, I didn't think that through.

Tested ACK b21912d455f225740fc5d4356cb4b31379c077d2

MeshCollider commented:
LGTM, light review utACK https://github.com/bitcoin/bitcoin/pull/11851/commits/b21912d455f225740fc5d4356cb4b31379c077d2

ryanofsky commented:
Rebased b21912d455f225740fc5d4356cb4b31379c077d2 -> 294442d56d0aff5571756980ee0e0f700656428e ([pr/wren.13](https://github.com/ryanofsky/bitcoin/commits/pr/wren.13) -> [pr/wren.14](https://github.com/ryanofsky/bitcoin/commits/pr/wren.14)) due to merge conflict with #12408.

jnewbery commented on src/wallet/walletdb.h:
```diff
@@ -244,8 +244,8 @@ class CWalletDB
 private:
-    CDB batch;
-    CWalletDBWrapper& m_dbw;
+    BerkeleyBatch batch;
```
You could rename this to `m_batch` in this PR, although that may be out of scope.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11851#discussion_r175532351

> You could rename this to m_batch in this PR, although that may be out of scope.

Done in c4719633496d6be338d3c2ee3d64be613200ae82

jnewbery reviewed (commented):
Tested ACK 294442d56d0aff5571756980ee0e0f700656428e. One nit inline, which could easily be addressed in a future PR.

ryanofsky committed:
Add m_ prefix to WalletBatch::m_batch

ryanofsky reviewed (commented):
Added 1 commits 294442d56d0aff5571756980ee0e0f700656428e -> c4719633496d6be338d3c2ee3d64be613200ae82 ([pr/wren.14](https://github.com/ryanofsky/bitcoin/commits/pr/wren.14) -> [pr/wren.15](https://github.com/ryanofsky/bitcoin/commits/pr/wren.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/wren.14...pr/wren.15))

jnewbery commented:
reACK c4719633496d6be338d3c2ee3d64be613200ae82 . Can confirm that the `batch` -> `m_batch` renaming is the only change.

MarcoFalke commented:
Needs rebase

ryanofsky commented:
> Needs rebase

Rebased c4719633496d6be338d3c2ee3d64be613200ae82 -> 2e05b7d4b6aa19adc30085af5cc2f6b050a7482d ([pr/wren.15](https://github.com/ryanofsky/bitcoin/commits/pr/wren.15) -> [pr/wren.16](https://github.com/ryanofsky/bitcoin/commits/pr/wren.16)) due to conflicts with #11536 and #12694.

Sjors commented:
Concept ACK on more clear variable names. No opinion on the chosen names, but they seem like an improvement. 4dff15c compiles for me on macOS and my wallets still work.

PierreRochard commented:
I should have caught this sooner, this comment https://github.com/ryanofsky/bitcoin/blob/70caf7f7538efe1f859ce849967b7b40bbb8ae6f/src/wallet/walletdb.h#L137 should be updated to reflect the renaming that happpened.

PierreRochard commented:
Tested ACK 9b0f0c5513f74825b06a728ab9bb2fd0e4224c9f

jnewbery commented:
Tested ACK 9b0f0c5513f74825b06a728ab9bb2fd0e4224c9f.

This is a conflict-heavy PR. It'd be great to get this merged soon to avoid repeated rebases.
